src/backend/distributed/utils/reference_table_utils.c

```diff
@@ -22,28 +22,331 @@
+	 * we don't want concurrent writes to pg_dist_node, we have already acquired
+	 * ShareLock on pg_dist_node.
+	 */
+	List *newWorkersList = WorkersWithoutReferenceTablePlacement(shardId,
+																 AccessShareLock);
+	if (list_length(newWorkersList) == 0)
+	{
+		/* nothing to do, no need for lock */
+		UnlockColocationId(colocationId, ExclusiveLock);
+		return;
+	}
+
+	/*
+	 * master_copy_shard_placement triggers metadata sync-up, which tries to
+	 * acquire a ShareLock on pg_dist_node. We do master_copy_shad_placement
+	 * in a separate connection. If we have modified pg_dist_node in the
+	 * current backend, this will cause a deadlock.
+	 */
+	if (TransactionModifiedNodeMetadata)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot replicate reference tables in a transaction "
+							   "that modified node metadata")));
+	}
+
+	/*
+	 * Modifications to reference tables in current transaction are not visible
+	 * to master_copy_shard_placement, since it is done in a separate backend.
+	 */
+	if (AnyRelationsModifiedInTransaction(referenceTableIdList))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot replicate reference tables in a transaction "
+							   "that modified a reference table")));
+	}
+
+	bool missingOk = false;
+	ShardPlacement *sourceShardPlacement = ActiveShardPlacement(shardId, missingOk);
+	if (sourceShardPlacement == NULL)
+	{
+		/* check for corrupt metadata */
+		ereport(ERROR, (errmsg("reference table shard "
+							   UINT64_FORMAT
+							   " does not have an active shard placement",
+							   shardId)));
+	}
+
+	/*
+	 * When using local backend for master_copy_shard_placement, current user should
+	 * own all reference table users. Otherwise master_copy_shard_placement will
+	 * error out. Superusers are regarded as owners.
+	 */
+	if (UseLocalBackendForReferenceTableReplication &&
+		!TableListOwner(referenceTableIdList))
+	{
+		ereport(ERROR, (errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+						errmsg("current user doesn't have enough permissions to "
+							   "replicate reference tables"),
+						errhint("Run 'SELECT replicate_reference_tables()' with "
+								"a superuser, and then try again.")));
+	}
+
+	WorkerNode *newWorkerNode = NULL;
+	foreach_ptr(newWorkerNode, newWorkersList)
+	{
+		ereport(NOTICE, (errmsg("replicating reference table '%s' to %s:%d ...",
+								referenceTableName, newWorkerNode->workerName,
+								newWorkerNode->workerPort)));
+
+		char transferMode = UseLocalBackendForReferenceTableReplication ?
+							TRANSFER_MODE_BLOCK_WRITES : TRANSFER_MODE_AUTOMATIC;
+
+		StringInfo placementCopyCommand =
+			CopyShardPlacementToWorkerNodeQuery(sourceShardPlacement,
+												newWorkerNode, transferMode);
+
+		if (UseLocalBackendForReferenceTableReplication)
+		{
+			ExecuteQueryStringIntoDestReceiver(placementCopyCommand->data,
+											   NULL, None_Receiver);
```

Comment:
Could we have a permissions issue here? We may need to use SetUserIdAndSecContext.